experimental/tools/handshake-simulator/ExecModels.cpp

```diff
@@ -162,207 +261,541 @@ LogicalResult dynamatic::experimental::initialiseMap(
+        continue;
+      }
+
+      if (request.lastExecution == data.currentCycle)
+       continue;
+
+      // Check if enough cycle passed (simulates the real circuit delay)
+      if (request.cyclesToComplete == 0) {
+        // Load the data accordingly
+        auto addressValue = data.valueMap[address];
+        unsigned offset = llvm::any_cast<APInt>(addressValue).getZExtValue();
+        auto &mem = data.store[bufferStart];
+        assert(offset < mem.size());
+        data.valueMap[dataOperand] = mem[offset];
+
+        mcData.loadRequests.erase(mcData.loadRequests.begin() + i);
+        hasDoneStuff = true;
+      } else {
+        --request.cyclesToComplete;
+      }
+      request.lastExecution = data.currentCycle;
+      data.internalDataMap[&opArg] = mcData;
+    }
+  }
+
+  return hasDoneStuff;
+}
+
+
+// NOTE PR : Honestly cant find a good way to simplify the repetitive Load
+//          and store code. They are very similar but differ from conditions,
+//          and my lambda versions didn't look any better idk
+bool DynamaticLoad::tryExecute(ExecutableData &data, circt::Operation &opArg) {
+  auto op = dyn_cast<circt::handshake::DynamaticLoadOp>(opArg);
+  bool hasDoneStuff = false;
+
+  // Send address to mem controller if available
+  if (!data.valueMap.count(op.getAddressResult())) {
+    memoryTransfer(op.getAddress(), op.getAddressResult(), data);
+    hasDoneStuff = true;
+  }
+  // Send data to successor if available
+  if (data.valueMap.count(op.getData()) && !data.valueMap.count(op.getDataResult())) {
+    memoryTransfer(op.getData(), op.getDataResult(), data);
+    hasDoneStuff = true;
+  }
+
+  return hasDoneStuff;
+}
+
+bool DynamaticStore::tryExecute(ExecutableData &data, circt::Operation &opArg) {
+  auto op = dyn_cast<circt::handshake::DynamaticStoreOp>(opArg);
+  bool hasDoneStuff = false;
+
+  // Send address to mem controller if available
+  if (data.valueMap.count(op.getAddress()) && !data.valueMap.count(op.getAddressResult())) {
+    memoryTransfer(op.getAddress(), op.getAddressResult(), data);
+    hasDoneStuff = true;
+  }
+  // Send data to mem controller if available
+  if (data.valueMap.count(op.getData()) && !data.valueMap.count(op.getDataResult())) {
+    memoryTransfer(op.getData(), op.getDataResult(), data);
+    hasDoneStuff = true;
+  }
+
+  return hasDoneStuff;
+}
+
+bool DynamaticReturn::tryExecute(ExecutableData &data,
+                                 circt::Operation &opArg) {
+  auto op = dyn_cast<circt::handshake::DynamaticReturnOp>(opArg);
+  auto executeFunc = [&](std::vector<llvm::Any> &ins,
+                         std::vector<llvm::Any> &outs, circt::Operation &op) {
+    for (unsigned i = 0; i < op.getNumOperands(); ++i)
+      outs[i] = ins[i];
+    data.internalDataMap[&op] = true;
+  };
+  return tryToExecute(op.getOperation(), data.valueMap, data.models,
+                      executeFunc);
+}
+
+bool DynamaticEnd::tryExecute(ExecutableData &data, circt::Operation &opArg) {
```

Comment:
I guess this is the first time I look seriously at this execution model. While what you have implemented may be equivalent in practice to the real component's behavior, it is not really the same and I think we should be somewhat accurate here.

We should consider the end of the execution reached when:
- all "return values" (which you can get with `endOp.getReturnValues`) are valid (i.e. are in the `valueMap` if I understand thing correctly). A return is basically a transparent buffer, meaning it more or less forwards its dataflow inputs to its dataflow outputs.
- all "memory done signals" (which you can get with `endOp.getMemoryControls()`) are valid. These are the last return values of all memory controllers.

Keep this for a future PR if it makes more sense, but keep it in mind.